sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java

```diff
@@ -0,0 +1,231 @@
+  public UnboundedSource.CheckpointMark getCheckpointMark() {
+    return checkpointMark;
+  }
+
+  @Override
+  public RabbitMqSource getCurrentSource() {
+    return source;
+  }
+
+  @Override
+  public byte[] getCurrentRecordId() {
+    return Optional.ofNullable(currentRecord).map(recordIdPolicy::apply).get();
+  }
+
+  @Override
+  public Instant getCurrentTimestamp() {
+    return Optional.ofNullable(currentRecord).map(timestampPolicy::getTimestampForRecord).get();
+  }
+
+  @Override
+  public RabbitMqMessage getCurrent() {
+    if (currentRecord == null) {
+      throw new NoSuchElementException();
+    }
+    return currentRecord;
+  }
+
+  @Override
+  public boolean start() throws IOException {
+    try {
+      checkpointMark.startReading();
+
+      ChannelLeaser.UseChannelFunction<String> setupFn =
+          channel -> {
+            ReadParadigm paradigm = source.spec.readParadigm();
+            String queueName = paradigm.queueName();
+            if (paradigm instanceof ReadParadigm.NewQueue) {
+              ReadParadigm.NewQueue newQueue = (ReadParadigm.NewQueue) paradigm;
+
+              final boolean durable = true;
+              final boolean autoDelete = false;
+              final boolean exclusive = false;
+              final Map<String, Object> arguments = Collections.emptyMap();
+
+              // this may fail if the queue isn't actually new and has different properties
+              channel.queueDeclare(queueName, durable, exclusive, autoDelete, arguments);
+
+              // the default exchange does not allow for explicitly binding a queue
+              if (!newQueue.isDefaultExchange()) {
+                channel.queueBind(queueName, newQueue.getExchange(), newQueue.getRoutingKey());
+              }
+            }
+
+            return queueName;
+          };
+
+      queueName = connectionHandler.useChannel(checkpointMark.getCheckpointId(), setupFn);
+    } catch (IOException e) {
+      checkpointMark.stopReading();
+      throw e;
+    } catch (Exception e) {
+      checkpointMark.stopReading();
+      throw new IOException(e);
+    }
+    return advance();
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    try {
+      // we consume message without autoAck (we want to do the ack ourselves)
+      GetResponse delivery =
+          connectionHandler.useChannel(
+              checkpointMark.getCheckpointId(),
+              channel -> channel.basicGet(queueName, /* autoAck: */ false));
+
+      if (delivery == null) {
+        currentRecord = null;
+        // queue is empty, so there is no backlog
+        context = new TimestampPolicyContext(0, Instant.now());
+        return false;
+      }
+
+      currentRecord = RabbitMqMessage.fromGetResponse(delivery);
+      // counting 'size' in terms of raw bytes of the body only
+      averageRecordSize.update(currentRecord.body().length);
+      context = new TimestampPolicyContext(delivery.getMessageCount(), Instant.now());
```

Comment:
`GetResponse` actually supplies an estimate of remaining queue backlog in terms of number of messages. Using that to produce an estimate of total backlog size.